File: src/test/java/com/agiledeveloper/MapsTest.java
package com.agiledeveloper;

import java.util.*;
import org.junit.jupiter.api.*;
import org.junit.jupiter.api.BeforeEach;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.platform.runner.JUnitPlatform;
import org.junit.runner.RunWith;

@RunWith(JUnitPlatform.class)
public class MapsTest {
  private Map<String, Integer> scores = new HashMap<String, Integer>() {
    {
      put("Jack", 12);
      put("Jill", 15);
      put("Tom", 11);
      put("Darla", 15);
      put("TOM", 11);
      put("Nick", 15);
      put("Nancy", 11);
    }
  };

  @Test
  void scoresList() {
    assertAll(
      () -> assertEquals(Arrays.asList("Jack"), Maps.listByScore(scores).get(12))
      , () -> assertEquals(3, Maps.listByScore(scores).get(15).size())
      , () -> assertTrue(Maps.listByScore(scores).get(15).contains("Darla"))
      , () -> assertEquals(3, Maps.listByScore(scores).get(11).size())
      , () -> assertTrue(Maps.listByScore(scores).get(11).contains("Tom"))
      , () -> assertTrue(Maps.listByScore(scores).get(11).contains("TOM"))
    );
  }

  @Test
  void scoresSet() {
    assertAll(
      () -> assertTrue(Maps.setByScore(scores).get(12).contains("JACK"))
      , () -> assertEquals(3, Maps.setByScore(scores).get(15).size())
      , () -> assertTrue(Maps.setByScore(scores).get(15).contains("DARLA"))
      , () -> assertEquals(2, Maps.setByScore(scores).get(11).size())
      , () -> assertTrue(Maps.setByScore(scores).get(11).contains("TOM"))
    );
  }

  @Test
  void scoresCount() {
    assertAll(
      () -> assertEquals(1, (int) Maps.countScores(scores).get(12))
      , () -> assertEquals(3, (int) Maps.countScores(scores).get(11))
      , () -> assertEquals(3, (int) Maps.countScores(scores).get(15))
    );
  }

  @Test
  void numberOfLetters() {
    assertAll(
      () -> assertTrue(Maps.numberOfLetters(scores).get(12).contains(4))
      , () -> assertTrue(Maps.numberOfLetters(scores).get(15).contains(5))
      , () -> assertTrue(Maps.numberOfLetters(scores).get(15).contains(4))
      , () -> assertTrue(Maps.numberOfLetters(scores).get(11).contains(5))
      , () -> assertTrue(Maps.numberOfLetters(scores).get(11).contains(3))
    );
  }

  @Test
  void maxNumberOfLetters() {
    assertAll(
      () -> assertEquals(4, (int) Maps.maxNumberOfLetters(scores).get(12))
      , () -> assertEquals(5, (int) Maps.maxNumberOfLetters(scores).get(15))
      , () -> assertEquals(5, (int) Maps.maxNumberOfLetters(scores).get(11))
    );
  }
}

File: src/test/scala/com/agiledeveloper/MapsTest.scala
package com.agiledeveloper

import org.scalatest._

class MapsTest extends WordSpec with Matchers {
  "The Maps functions" when {
    "given a scores map" should {
      val scores = Map(
        "Jack" -> 12,
        "Jill" -> 15,
        "Tom" -> 11,
        "Darla" -> 15,
        "TOM" -> 11,
        "Nick" -> 15,
        "Nancy" -> 11
      )

      "collect names by scores" in {
        Maps.listByScore(scores)(12) shouldBe List("Jack")
        Maps.listByScore(scores)(15) should have size 3
        Maps.listByScore(scores)(15) should contain ("Darla")
        Maps.listByScore(scores)(11) should have size 3
        Maps.listByScore(scores)(11) should contain ("Tom")
        Maps.listByScore(scores)(11) should contain ("TOM")
      }

      "collect names sets by score" in {
        Maps.setByScore(scores)(12) should contain ("JACK")
        Maps.setByScore(scores)(15) should have size 3
        Maps.setByScore(scores)(15) should contain ("DARLA")
        Maps.setByScore(scores)(11) should have size 2
        Maps.setByScore(scores)(11) should contain ("TOM")
      }

      "count scores" in {
        Maps.countScores(scores)(12) shouldBe 1
        Maps.countScores(scores)(11) shouldBe 3
        Maps.countScores(scores)(15) shouldBe 3
      }

      "count letters for each score" in {
        Maps.numberOfLetters(scores)(12) should contain (4)
        Maps.numberOfLetters(scores)(15) should contain (5)
        Maps.numberOfLetters(scores)(15) should contain (4)
        Maps.numberOfLetters(scores)(11) should contain (5)
        Maps.numberOfLetters(scores)(11) should contain (3)
      }

      "compute the longest name for each score" in {
        Maps.maxNumberOfLetters(scores)(12) shouldBe 4
        Maps.maxNumberOfLetters(scores)(15) shouldBe 5
        Maps.maxNumberOfLetters(scores)(11) shouldBe 5
      }
    }
  }
}
